File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/isolationpayment/IsolationPaymentPersistence.java
package uk.nhs.nhsx.isolationpayment;

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB;
import com.amazonaws.services.dynamodbv2.model.*;
import uk.nhs.nhsx.isolationpayment.model.IsolationToken;
import uk.nhs.nhsx.isolationpayment.model.TokenStateInternal;

import java.util.Map;
import java.util.Optional;

import static uk.nhs.nhsx.core.aws.dynamodb.DynamoAttributes.*;

public class IsolationPaymentPersistence {

    private final AmazonDynamoDB dynamoDbClient;
    private final String tableName;

    public IsolationPaymentPersistence(AmazonDynamoDB dynamoDbClient, String tableName) {
        this.dynamoDbClient = dynamoDbClient;
        this.tableName = tableName;
    }

    public void insertIsolationToken(IsolationToken token) {
        var request = new PutItemRequest()
            .withTableName(tableName)
            .withItem(asAttributes(token));

        dynamoDbClient.putItem(request);
    }

    public void updateIsolationToken(IsolationToken token, TokenStateInternal currentTokenStatus) {
        var request = new PutItemRequest()
            .withTableName(tableName)
            .withItem(asAttributes(token))
            .withConditionExpression("attribute_exists(tokenId) and tokenStatus=:tokenStatus")
            .withExpressionAttributeValues(Map.of(":tokenStatus", stringAttribute(currentTokenStatus.value)));

        dynamoDbClient.putItem(request);
    }

    private Map<String, AttributeValue> asAttributes(IsolationToken token) {
        return Map.of(
            "tokenId", stringAttribute(token.tokenId),
            "tokenStatus", stringAttribute(token.tokenStatus),
            "riskyEncounterDate", numericNullableAttribute(token.riskyEncounterDate),
            "isolationPeriodEndDate", numericNullableAttribute(token.isolationPeriodEndDate),
            "createdTimestamp", numericAttribute(token.createdTimestamp),
            "updatedTimestamp", numericNullableAttribute(token.updatedTimestamp),
            "validatedTimestamp", numericNullableAttribute(token.validatedTimestamp),
            "consumedTimestamp", numericNullableAttribute(token.consumedTimestamp),
            "expireAt", numericAttribute(token.expireAt)
        );
    }

    public Optional<IsolationToken> getIsolationToken(String ipcToken) {
        var request = new GetItemRequest()
            .withTableName(tableName)
            .withKey(Map.of("tokenId", stringAttribute(ipcToken)));
        
        var itemResult = dynamoDbClient.getItem(request);

        return Optional.ofNullable(itemResult.getItem())
            .map(it -> new IsolationToken(
                itemValueOrThrow(it, "tokenId"),
                itemValueOrThrow(it, "tokenStatus"),
                itemLongValueMaybe(it, "riskyEncounterDate").orElse(null),
                itemLongValueMaybe(it, "isolationPeriodEndDate").orElse(null),
                itemLongValueMaybe(it, "createdTimestamp").orElse(null),
                itemLongValueMaybe(it, "updatedTimestamp").orElse(null),
                itemLongValueMaybe(it, "validatedTimestamp").orElse(null),
                itemLongValueMaybe(it, "consumedTimestamp").orElse(null),
                itemLongValueMaybe(it, "expireAt").orElse(null)
            ));
    }

    public void deleteIsolationToken(String ipcToken, TokenStateInternal currentTokenStatus) {
        var request = new DeleteItemRequest()
            .withTableName(tableName)
            .withKey(Map.of("tokenId", stringAttribute(ipcToken)))
            .withConditionExpression("attribute_exists(tokenId) and tokenStatus=:tokenStatus")
            .withExpressionAttributeValues(Map.of(":tokenStatus", stringAttribute(currentTokenStatus.value)));

        dynamoDbClient.deleteItem(request);
    }
}

File: src/aws/lambdas/incremental_distribution/src/main/java/uk/nhs/nhsx/isolationpayment/IsolationPaymentPersistence.kt
package uk.nhs.nhsx.isolationpayment

import com.amazonaws.services.dynamodbv2.AmazonDynamoDB
import com.amazonaws.services.dynamodbv2.model.AttributeValue
import com.amazonaws.services.dynamodbv2.model.DeleteItemRequest
import com.amazonaws.services.dynamodbv2.model.GetItemRequest
import com.amazonaws.services.dynamodbv2.model.PutItemRequest
import uk.nhs.nhsx.core.aws.dynamodb.DynamoAttributes.itemLongValueOrNull
import uk.nhs.nhsx.core.aws.dynamodb.DynamoAttributes.itemLongValueOrThrow
import uk.nhs.nhsx.core.aws.dynamodb.DynamoAttributes.itemValueOrThrow
import uk.nhs.nhsx.core.aws.dynamodb.DynamoAttributes.numericAttribute
import uk.nhs.nhsx.core.aws.dynamodb.DynamoAttributes.numericNullableAttribute
import uk.nhs.nhsx.core.aws.dynamodb.DynamoAttributes.stringAttribute
import uk.nhs.nhsx.isolationpayment.model.IsolationToken
import uk.nhs.nhsx.isolationpayment.model.TokenStateInternal
import uk.nhs.nhsx.domain.IpcTokenId
import java.util.Optional

class IsolationPaymentPersistence(private val dynamoDbClient: AmazonDynamoDB, private val tableName: String) {
    fun insertIsolationToken(token: IsolationToken) {
        dynamoDbClient.putItem(
            PutItemRequest()
                .withTableName(tableName)
                .withItem(asAttributes(token))
        )
    }

    fun updateIsolationToken(token: IsolationToken, currentTokenStatus: TokenStateInternal) {
        dynamoDbClient.putItem(
            PutItemRequest()
                .withTableName(tableName)
                .withItem(asAttributes(token))
                .withConditionExpression("attribute_exists(tokenId) and tokenStatus=:tokenStatus")
                .withExpressionAttributeValues(
                    mapOf(
                        ":tokenStatus" to stringAttribute(currentTokenStatus.value)
                    )
                )
        )
    }

    private fun asAttributes(token: IsolationToken): Map<String, AttributeValue> = mapOf(
        "tokenId" to stringAttribute(token.tokenId),
        "tokenStatus" to stringAttribute(token.tokenStatus),
        "riskyEncounterDate" to numericNullableAttribute(token.riskyEncounterDate),
        "isolationPeriodEndDate" to numericNullableAttribute(token.isolationPeriodEndDate),
        "createdTimestamp" to numericAttribute(token.createdTimestamp),
        "updatedTimestamp" to numericNullableAttribute(token.updatedTimestamp),
        "validatedTimestamp" to numericNullableAttribute(token.validatedTimestamp),
        "consumedTimestamp" to numericNullableAttribute(token.consumedTimestamp),
        "expireAt" to numericAttribute(token.expireAt)
    )

    fun getIsolationToken(ipcToken: IpcTokenId): Optional<IsolationToken> {
        val request = GetItemRequest()
            .withTableName(tableName)
            .withKey(mapOf("tokenId" to stringAttribute(ipcToken)))
        val itemResult = dynamoDbClient.getItem(request)

        return Optional.ofNullable(itemResult.item?.let {
            IsolationToken(
                IpcTokenId.of(itemValueOrThrow(it, "tokenId")),
                itemValueOrThrow(it, "tokenStatus"),
                itemLongValueOrNull(it, "riskyEncounterDate"),
                itemLongValueOrNull(it, "isolationPeriodEndDate"),
                itemLongValueOrThrow(it, "createdTimestamp"),
                itemLongValueOrNull(it, "updatedTimestamp"),
                itemLongValueOrNull(it, "validatedTimestamp"),
                itemLongValueOrNull(it, "consumedTimestamp"),
                itemLongValueOrThrow(it, "expireAt")
            )
        })
    }

    fun deleteIsolationToken(ipcToken: IpcTokenId, currentTokenStatus: TokenStateInternal) {
        dynamoDbClient.deleteItem(
            DeleteItemRequest()
                .withTableName(tableName)
                .withKey(java.util.Map.of("tokenId", stringAttribute(ipcToken)))
                .withConditionExpression("attribute_exists(tokenId) and tokenStatus=:tokenStatus")
                .withExpressionAttributeValues(mapOf(":tokenStatus" to stringAttribute(currentTokenStatus.value)))
        )
    }
}
